### PR TITLE
Enhancement: Improvements to iohyve list and info

### PR DESCRIPTION
**Pull request by EpiJunkie**
This pull request contains three commits to improve the visibility of multiple pool installations. See each commit for full details.

2f2b3c7:
`iohyve info` was rewritten to include the description when spaces are present. Each pool gets it's own table of output but everything is aligned as a single block of output for readability. The `-d` flag was removed.

0548ceb
`iohyve list` now contains the pool name in the output. This is to increase visibility for multi pool installations.

1f44663
This supports the documentation (man page, text, and `iohyve help`) to reflect changes from 2f2b3c7.

### CODE REVIEW

**Comment by pr1ntf** [comment]
Alright, this is a tricky one. 

I already had some list cleaning I wanted to do (briefly mentioned [here](https://github.com/pr1ntf/iohyve/wiki/General-Path-to-1.0#09-goals-submit-to-ports)).
Part of this was completely removing `description` from the `info` function and adding `tap` and `con` instead.
I would like to keep the `list` function simple, with basically what it has now, so maybe adding `pool` to the `info` function would be better. 

Last but not least, I wanted to add a `--less` flag so you could scroll through a large list that takes up too much of the terminal. 

If you do not want to make these changes, no problem, I'll add these specific points to the roadmap above. Especially the `pool` option. 

I'm open to discussion.

**Comment by pr1ntf** [comment]
Also, I'd like to mention we generate the `iohyve.8` man page with a one-liner utilizing `txt2man`:

```
cat iohyve.8.txt | txt2man -t iohyve -s 8 -v "FreeBSD System Manager's Manual" > iohyve.8
```

**Comment by EpiJunkie** [comment]
Thank you for the heads up on the rebuilding the `man` page file. I noticed the header about the same time you posted and was trying to figure it out when I noticed your message here. Thank you.

Added commits to update a couple definitions and then a properly rebuilt `man` page file.

**Comment by pr1ntf** [comment]
Did you happen to see my previous notes before the man page comment?

**Comment by EpiJunkie** [comment]
I missed that post actually, thanks.

First thing I _have to_ ask, is do you really have that many guests that `--less` is necessary? Can you ballpark a number?

What if I kept the output format I have created so far but put it under `-v` for verbose and then made something more to your liking under the normal command? Also maybe use `-l` instead of `--less`.

**Comment by pr1ntf** [comment]
```
tparis@DeltaFlyer:~/iohyve % iohyve list | wc -l
      46
tparis@DeltaFlyer:~/iohyve % 
```

They all don't run at the same time, only about ~8ish at a time, but I have different VMs for different projects and ideas. Some are clones of templates, too.  I usually just manually run `iohyve list | less` or `iohyve info | less`.  You should see my `snaplist` output. 

Let's keep the `list` function untouched, but remove `description` from the `info` function. 

Maybe info without options will output `size`, `cpu`, `ram`, and `pool`.  
Info with the `-v` flag can add `os`, `loader`, `tap`, and `con`.

EDIT: Didn't mean to close, fingers got crossed.

**Comment by pr1ntf** [comment]
Be sure to read the previous post on GitHub, not on email, so you can see my edits before I accidentally closed this PR.

**Comment by EpiJunkie** [comment]
Wow, that's quiet a few guests.

I will work on these changes this weekend.

**Comment by allenmchan** [comment]
Is it possible to list IPs of the VMs when using the iohyve list command?
I have all of the VMs using DHCP. It is a pain to console into each VM to get the IP

**Comment by EpiJunkie** [comment]
Not within `iohyve`, especially when using DHCP. Maybe add static DHCP mapping to your DHCP server and then add the IP in the `description` property?

**Comment by allenmchan** [comment]
Ok thanks for the workaround

**Comment by EpiJunkie** [comment]
Okay I made some changes. I hope they are to your liking.

There is more code in `list()` _but_ there are flags to dial up the verbosity. The description field is removed by default as requested above. The tap and con fields are turned on by `-v`, the description field is turned back on by `-d`flag, and using the `-l` flag pipes the output to `less`. 

I also reworked `info()` as well, the code is very similar to `list()`. There is a lot more code due to the extra fields but I am very happy with the flexibility of it. The old code was so concise. This sub-function follows the same pattern as `list` where the verbosity can be dialed up using flags. The default output is to list guest name, size, CPU, and RAM. Using the `-p` flag turns on the pool field; the `-v`(verbose) flag turns on the OS, loader, tap, and con fields; the `-s` (status) flag turns on the VMM, running, and rcboot fields; the `-d` flag turns on the description field; and using the `-l` flag pipes the output to `less`.

To use some of the flags for `info` the command would be `iohyve info -psl` or any order on the flags. The order of the flags does not influence the order of the columns, sorry. Any combination of flags can be used, I wrote the code to be flexible enough for that.

Let me know if you would like to see any further changes. I hope you like the results.

Also @pr1ntf thank you for amusing my curiosity with the number of guests. It was pure curiosity, I re-read the question and realized it could of come off as aggressive. My sincerest apologies if it came off that way, not my intention at all. Also thank you with the follow ups yesterday, it was quiet a long day with little sleep.

## Here is the before and after for `iohyve list`

**Before:**

```
root@bhost:~ # iohyve list
Guest      VMM?  Running  rcboot?  Description
testing    YES   YES      NO       Thu Apr  7 10:37:59 MDT 2016
guest      YES   YES      NO       Ubiquiti mFi Controller Appliance
gst01      NO    NO       NO       Internal https
obsdguest  NO    NO       NO       Wed Apr  6 14:18:44 MDT 2016
```

**After:**

```
root@bhost:~ # iohyve list
Guest      VMM?  Running  rcboot?
testing    YES   YES      NO
guest      YES   YES      NO
gst01      NO    NO       NO
obsdguest  NO    NO       NO
root@bhost:~ # iohyve list -d
Guest      VMM?  Running  rcboot?  Description
testing    YES   YES      NO       Thu Apr  7 10:37:59 MDT 2016
guest      YES   YES      NO       Ubiquiti mFi Controller Appliance
gst01      NO    NO       NO       Internal https
obsdguest  NO    NO       NO       Wed Apr  6 14:18:44 MDT 2016
root@bhost:~ # iohyve list -dv
Guest      VMM?  Running  rcboot?  Tap      Con     Description
testing    YES   YES      NO       tap7011  nmdm17  Thu Apr  7 10:37:59 MDT 2016
guest      YES   YES      NO       tap7003  nmdm1   Ubiquiti mFi Controller Appliance
gst01      NO    NO       NO       tap7004  nmdm10  Internal https
obsdguest  NO    NO       NO       tap7010  nmdm16  Wed Apr  6 14:18:44 MDT 2016
root@bhost:~ # iohyve list -dvl
root@bhost:~ # 
```

## Here is the before and after for `iohyve info`

**Before:**

```
root@bhost:~ # iohyve info
Name             Size  RAM     CPU  OS         Loader
testing          -     256M    1    default    bhyveload
testing/disk0    10G   256M    1    default    bhyveload
guest            -     4G      2    debian     grub-bhyve
guest/disk0      200G  4G      2    debian     grub-bhyve
gst01            -     512M    1    default    bhyveload
gst01/disk0      40G   512M    1    default    bhyveload
obsdguest        -     256M    1    openbsd59  grub-bhyve
obsdguest/disk0  8G    256M    1    openbsd59  grub-bhyve
root@bhost:~ # iohyve info -d
Name             Size  RAM     CPU  OS         Loader      Description
testing          -     256M    1    default    bhyveload   Thu          Apr  7           10:37:59          MDT      2016
testing/disk0    10G   256M    1    default    bhyveload   Thu          Apr  7           10:37:59          MDT      2016
guest            -     4G      2    debian     grub-bhyve  Ubiquiti     mFi  Controller  Appliance
guest/disk0      200G  4G      2    debian     grub-bhyve  Ubiquiti     mFi  Controller  Appliance
gst01            -     512M    1    default    bhyveload   Internal         https
gst01/disk0      40G   512M    1    default    bhyveload   Internal         https
obsdguest        -     256M    1    openbsd59  grub-bhyve  Wed          Apr  6           14:18:44          MDT      2016
obsdguest/disk0  8G    256M    1    openbsd59  grub-bhyve  Wed          Apr  6           14:18:44          MDT      2016
```

**After:**

```
root@bhost:~ # iohyve info             
Guest            Size  CPU  RAM          
testing          -     1    256M         
testing/disk0    10G   1    256M         
guest            -     2    4G           
guest/disk0      200G  2    4G           
gst01            -     1    512M         
gst01/disk0      40G   1    512M         
obsdguest        -     1    256M         
obsdguest/disk0  8G    1    256M         
root@bhost:~ # iohyve info -p          
Guest            Size  CPU  RAM     Pool 
testing          -     1    256M    apool          
testing/disk0    10G   1    256M    apool          
guest            -     2    4G      zpool          
guest/disk0      200G  2    4G      zpool          
gst01            -     1    512M    zpool          
gst01/disk0      40G   1    512M    zpool          
obsdguest        -     1    256M    zpool          
obsdguest/disk0  8G    1    256M    zpool          
root@bhost:~ # iohyve info -pv         
Guest            Size  CPU  RAM     Pool       OS         Loader      Tap      Con                 
testing          -     1    256M    apool      default    bhyveload   tap7011  nmdm17              
testing/disk0    10G   1    256M    apool      default    bhyveload   tap7011  nmdm17              
guest            -     2    4G      zpool      debian     grub-bhyve  tap7003  nmdm1               
guest/disk0      200G  2    4G      zpool      debian     grub-bhyve  tap7003  nmdm1               
gst01            -     1    512M    zpool      default    bhyveload   tap7004  nmdm10              
gst01/disk0      40G   1    512M    zpool      default    bhyveload   tap7004  nmdm10              
obsdguest        -     1    256M    zpool      openbsd59  grub-bhyve  tap7010  nmdm16              
obsdguest/disk0  8G    1    256M    zpool      openbsd59  grub-bhyve  tap7010  nmdm16              
root@bhost:~ # iohyve info -pvs        
Guest            Size  CPU  RAM     Pool       OS         Loader      Tap      Con     VMM?  Running  rcboot?
testing          -     1    256M    apool      default    bhyveload   tap7011  nmdm17  YES   YES      NO     
testing/disk0    10G   1    256M    apool      default    bhyveload   tap7011  nmdm17  NO    NO       NO     
guest            -     2    4G      zpool      debian     grub-bhyve  tap7003  nmdm1   NO    YES      NO     
guest/disk0      200G  2    4G      zpool      debian     grub-bhyve  tap7003  nmdm1   NO    NO       NO     
gst01            -     1    512M    zpool      default    bhyveload   tap7004  nmdm10  NO    NO       NO     
gst01/disk0      40G   1    512M    zpool      default    bhyveload   tap7004  nmdm10  NO    NO       NO     
obsdguest        -     1    256M    zpool      openbsd59  grub-bhyve  tap7010  nmdm16  NO    NO       NO     
obsdguest/disk0  8G    1    256M    zpool      openbsd59  grub-bhyve  tap7010  nmdm16  NO    NO       NO     
root@bhost:~ # iohyve info -pvsd       
Guest            Size  CPU  RAM     Pool       OS         Loader      Tap      Con     VMM?  Running  rcboot?  Description      
testing          -     1    256M    apool      default    bhyveload   tap7011  nmdm17  YES   YES      NO       Thu Apr  7 10:37:59 MDT 2016                  
testing/disk0    10G   1    256M    apool      default    bhyveload   tap7011  nmdm17  NO    NO       NO       Thu Apr  7 10:37:59 MDT 2016                  
guest            -     2    4G      zpool      debian     grub-bhyve  tap7003  nmdm1   NO    YES      NO       Ubiquiti mFi Controller Appliance            
guest/disk0      200G  2    4G      zpool      debian     grub-bhyve  tap7003  nmdm1   NO    NO       NO       Ubiquiti mFi Controller Appliance            
gst01            -     1    512M    zpool      default    bhyveload   tap7004  nmdm10  NO    NO       NO       Internal https            
gst01/disk0      40G   1    512M    zpool      default    bhyveload   tap7004  nmdm10  NO    NO       NO       Internal https            
obsdguest        -     1    256M    zpool      openbsd59  grub-bhyve  tap7010  nmdm16  NO    NO       NO       Wed Apr  6 14:18:44 MDT 2016                  
obsdguest/disk0  8G    1    256M    zpool      openbsd59  grub-bhyve  tap7010  nmdm16  NO    NO       NO       Wed Apr  6 14:18:44 MDT 2016                  
root@bhost:~ # iohyve info -pvsdl
root@bhost:~ # 
```

File: iohyve [hunk 1]
```diff
@@ -258,9 +258,25 @@ __get_zfs_pcidev_conf() {
 
 # List Guests
 __list() {
+	local flags="$1"
 	local pools="$(zfs list -H | grep iohyve | cut -d '/' -f 1 | uniq )"
 	(
-        printf "%s^%s^%s^%s^%s\n" "Guest" "VMM?" "Running" "rcboot?" "Description"
+
```
Comment (EpiJunkie) [inline iohyve]: Using this method, the flags are al-la-carte allowing for flexibility and not using as much code as if you were to create an if statement or sub-sub-function for each desired output.

File: iohyve [hunk 2]
```diff
@@ -282,27 +298,144 @@ __list() {
...
+		else
+			cat   
+		fi
+	)
+
 }
 
 # Display info about all guests.
 __info() {
-	dflag="$2"
```
Comment (EpiJunkie) [inline iohyve]: I really do like this bit of code, seriously, so concise with lots of formatted output.

File: iohyve [hunk 3]
```diff
@@ -282,27 +298,144 @@ __list() {
...
+		if [ -n "$(echo "$flags" | grep 'd')" ]; then
+			local description="$(zfs get -H -o value iohyve:description $pool/iohyve/$g)"
+			printf "^%s" "$description"
+		fi
+		
+		printf "\n" #Print new line for next guest.
             done
 	done
-	) | column -ts^
+	) | column -ts^ | \
```
Comment (EpiJunkie) [inline iohyve]: This was inspired by the function above piping to `column`. I figured the same could be done for a smaller bit of code needing to get piped.

File: iohyve [hunk 4]
```diff
@@ -258,9 +258,25 @@ __get_zfs_pcidev_conf() {
...
+	if [ -n "$(echo "$flags" | grep 'v')" ]; then
+		printf "^%s^%s" "Tap" "Con"
+	fi
+	
+	# Print headers for -d flag (description)
+	if [ -n "$(echo "$flags" | grep 'd')" ]; then
+		printf "^%s" "Description"
+	fi
+	
+	printf "\n" # Print new line for first guest.
```
Comment (EpiJunkie) [inline iohyve]: Using this method, the flags can be used al-la-carte without needing to write static blocks of code for the desired output.

**Comment by pr1ntf** [comment]
Great work! Buuuuuuut..... I'm sorry for the confusion, I wanted the `list` function to remain unchanged. 

> Let's keep the list function untouched, but remove description from the info function.
> 
> Maybe info without options will output size, cpu, ram, and pool.
> 
> Info with the -v flag can add os, loader, tap, and con.

I'm so sorry to be the benevolent dictator of iohyve, especially when you've been helping out so much. 
If you've got other ideas and suggestions, I'm all ears.

**Comment by EpiJunkie** [comment]
Is it alright if the other flags remain? 

If I am understanding correctly, below is the desired output for flagless and using `-v` flags. The `-l` is there as a request to pipe to `less`. The other flags are there because they make my heart go pitter-pat. In addition, if new properties are added with this code base, they will be easy to flag in without changing the look from previous versions.

### Here is the before and after for `iohyve list` (Revised)

**Before:**

```
root@bhost:~ # iohyve list
Guest      VMM?  Running  rcboot?  Description
testing    YES   YES      NO       Thu Apr  7 10:37:59 MDT 2016
guest      YES   YES      NO       Ubiquiti mFi Controller Appliance
gst01      NO    NO       NO       Internal https
obsdguest  NO    NO       NO       Wed Apr  6 14:18:44 MDT 2016
```

**After:**

```
root@bhost:~ # iohyve list
Guest      VMM?  Running  rcboot?  Description
testing    YES   YES      NO       Thu Apr  7 10:37:59 MDT 2016
guest      YES   YES      NO       Ubiquiti mFi Controller Appliance
gst01      NO    NO       NO       Internal https
obsdguest  NO    NO       NO       Wed Apr  6 14:18:44 MDT 2016
root@bhost:~ # iohyve list -v
Guest      VMM?  Running  rcboot?  Tap      Con     Description
testing    YES   YES      NO       tap7011  nmdm17  Thu Apr  7 10:37:59 MDT 2016
guest      YES   YES      NO       tap7003  nmdm1   Ubiquiti mFi Controller Appliance
gst01      NO    NO       NO       tap7004  nmdm10  Internal https
obsdguest  NO    NO       NO       tap7010  nmdm16  Wed Apr  6 14:18:44 MDT 2016
root@bhost:~ # iohyve list -vl
root@bhost:~ # 
```

### Here is the before and after for `iohyve info` (Revised)

**Before:**

```
root@bhost:~ # iohyve info
Name             Size  RAM     CPU  OS         Loader
testing          -     256M    1    default    bhyveload
testing/disk0    10G   256M    1    default    bhyveload
guest            -     4G      2    debian     grub-bhyve
guest/disk0      200G  4G      2    debian     grub-bhyve
gst01            -     512M    1    default    bhyveload
gst01/disk0      40G   512M    1    default    bhyveload
obsdguest        -     256M    1    openbsd59  grub-bhyve
obsdguest/disk0  8G    256M    1    openbsd59  grub-bhyve
root@bhost:~ # iohyve info -d
Name             Size  RAM     CPU  OS         Loader      Description
testing          -     256M    1    default    bhyveload   Thu          Apr  7           10:37:59          MDT      2016
testing/disk0    10G   256M    1    default    bhyveload   Thu          Apr  7           10:37:59          MDT      2016
guest            -     4G      2    debian     grub-bhyve  Ubiquiti     mFi  Controller  Appliance
guest/disk0      200G  4G      2    debian     grub-bhyve  Ubiquiti     mFi  Controller  Appliance
gst01            -     512M    1    default    bhyveload   Internal         https
gst01/disk0      40G   512M    1    default    bhyveload   Internal         https
obsdguest        -     256M    1    openbsd59  grub-bhyve  Wed          Apr  6           14:18:44          MDT      2016
obsdguest/disk0  8G    256M    1    openbsd59  grub-bhyve  Wed          Apr  6           14:18:44          MDT      2016
```

**After:**

```
root@bhost:~ # iohyve info                
Guest            Size  CPU  RAM     Pool 
testing          -     1    256M    apool          
testing/disk0    10G   1    256M    apool          
guest            -     2    4G      zpool          
guest/disk0      200G  2    4G      zpool          
gst01            -     1    512M    zpool          
gst01/disk0      40G   1    512M    zpool          
obsdguest        -     1    256M    zpool          
obsdguest/disk0  8G    1    256M    zpool          
root@bhost:~ # iohyve info -v         
Guest            Size  CPU  RAM     Pool       OS         Loader      Tap      Con                 
testing          -     1    256M    apool      default    bhyveload   tap7011  nmdm17              
testing/disk0    10G   1    256M    apool      default    bhyveload   tap7011  nmdm17              
guest            -     2    4G      zpool      debian     grub-bhyve  tap7003  nmdm1               
guest/disk0      200G  2    4G      zpool      debian     grub-bhyve  tap7003  nmdm1               
gst01            -     1    512M    zpool      default    bhyveload   tap7004  nmdm10              
gst01/disk0      40G   1    512M    zpool      default    bhyveload   tap7004  nmdm10              
obsdguest        -     1    256M    zpool      openbsd59  grub-bhyve  tap7010  nmdm16              
obsdguest/disk0  8G    1    256M    zpool      openbsd59  grub-bhyve  tap7010  nmdm16              
root@bhost:~ # iohyve info -vs        
Guest            Size  CPU  RAM     Pool       OS         Loader      Tap      Con     VMM?  Running  rcboot?
testing          -     1    256M    apool      default    bhyveload   tap7011  nmdm17  YES   YES      NO     
testing/disk0    10G   1    256M    apool      default    bhyveload   tap7011  nmdm17  NO    NO       NO     
guest            -     2    4G      zpool      debian     grub-bhyve  tap7003  nmdm1   NO    YES      NO     
guest/disk0      200G  2    4G      zpool      debian     grub-bhyve  tap7003  nmdm1   NO    NO       NO     
gst01            -     1    512M    zpool      default    bhyveload   tap7004  nmdm10  NO    NO       NO     
gst01/disk0      40G   1    512M    zpool      default    bhyveload   tap7004  nmdm10  NO    NO       NO     
obsdguest        -     1    256M    zpool      openbsd59  grub-bhyve  tap7010  nmdm16  NO    NO       NO     
obsdguest/disk0  8G    1    256M    zpool      openbsd59  grub-bhyve  tap7010  nmdm16  NO    NO       NO     
root@bhost:~ # iohyve info -vsd       
Guest            Size  CPU  RAM     Pool       OS         Loader      Tap      Con     VMM?  Running  rcboot?  Description      
testing          -     1    256M    apool      default    bhyveload   tap7011  nmdm17  YES   YES      NO       Thu Apr  7 10:37:59 MDT 2016                  
testing/disk0    10G   1    256M    apool      default    bhyveload   tap7011  nmdm17  NO    NO       NO       Thu Apr  7 10:37:59 MDT 2016                  
guest            -     2    4G      zpool      debian     grub-bhyve  tap7003  nmdm1   NO    YES      NO       Ubiquiti mFi Controller Appliance            
guest/disk0      200G  2    4G      zpool      debian     grub-bhyve  tap7003  nmdm1   NO    NO       NO       Ubiquiti mFi Controller Appliance            
gst01            -     1    512M    zpool      default    bhyveload   tap7004  nmdm10  NO    NO       NO       Internal https            
gst01/disk0      40G   1    512M    zpool      default    bhyveload   tap7004  nmdm10  NO    NO       NO       Internal https            
obsdguest        -     1    256M    zpool      openbsd59  grub-bhyve  tap7010  nmdm16  NO    NO       NO       Wed Apr  6 14:18:44 MDT 2016                  
obsdguest/disk0  8G    1    256M    zpool      openbsd59  grub-bhyve  tap7010  nmdm16  NO    NO       NO       Wed Apr  6 14:18:44 MDT 2016                  
root@bhost:~ # iohyve info -vsdl
root@bhost:~ # 
```

**Comment by pr1ntf** [comment]
Don't even worry about touching the `description` function at all. 

The info changes look great, though! :+1: 

Again, sorry for the confusion. 

I've done quite a bit to the script the past week, there are conflicts with this PR.

**Comment by EpiJunkie** [comment]
I am guessing you meant `list` function. I left the `-l` flag to pipe to less. Let me know if that needs to be removed too.

Thanks.

**Comment by pr1ntf** [comment]
Lookin good in the neighborhood.

**Pull request merged**